compiler/hash-typecheck/src/ops/unify.rs

```diff
@@ -278,6 +284,44 @@ impl<'gs, 'ls, 'cd, 's> Unifier<'gs, 'ls, 'cd, 's> {
                 }
             }
 
+            // Union:
+            (_, Term::Union(inner_target)) => {
+                // Try to merge each individual term in source, with target. If any one
+                // succeeds, then the whole thing should succeed.
+                let mut first_error = None;
+                for inner_target_id in inner_target {
+                    match self.unify_terms(simplified_src_id, inner_target_id) {
+                        Ok(result) => {
+                            return Ok(result);
+                        }
+                        Err(e) if first_error.is_none() => {
+                            first_error = Some(e);
+                            continue;
+                        }
+                        _ => continue,
```

Comment:
maybe use `Option::or` here? It seems to be a nice fit for this, [this](https://doc.rust-lang.org/std/option/enum.Option.html#method.or) is the reference to the API